iroh-net/src/hp/netcheck/reportgen.rs

```diff
@@ -0,0 +1,993 @@
+
+        // Once we've heard from enough regions (3), start a timer to
+        // give up on the other ones. The timer's duration is a
+        // function of whether this is our initial full probe or an
+        // incremental one. For incremental ones, wait for the
+        // duration of the slowest region. For initial ones, double that.
+        if self.report.region_latency.len() == ENOUGH_REGIONS {
+            let mut timeout = self.report.region_latency.max_latency();
+            if !self.incremental {
+                timeout *= 2;
+            }
+            let reportcheck = self.addr();
+            tokio::spawn(async move {
+                time::sleep(timeout).await;
+                reportcheck.send(Message::AbortProbes).await.ok();
+            });
+        }
+
+        if let Some(ipp) = ipp {
+            match ipp {
+                SocketAddr::V4(_) => {
+                    self.report
+                        .region_v4_latency
+                        .update_region(node.region_id, latency);
+                    self.report.ipv4 = true;
+                    if self.report.global_v4.is_none() {
+                        self.report.global_v4 = Some(ipp);
+                    } else if self.report.global_v4 != Some(ipp) {
+                        self.report.mapping_varies_by_dest_ip = Some(true);
+                    } else if self.report.mapping_varies_by_dest_ip.is_none() {
+                        self.report.mapping_varies_by_dest_ip = Some(false);
+                    }
+                }
+                SocketAddr::V6(_) => {
+                    self.report
+                        .region_v6_latency
+                        .update_region(node.region_id, latency);
+                    self.report.ipv6 = true;
+                    self.report.global_v6 = Some(ipp);
+                    // TODO: track MappingVariesByDestIP for IPv6 too? Would be sad if so, but
+                    // who knows.
+                }
+            }
+        }
+    }
+
+    /// Stops further probes.
+    ///
+    /// This makes sure that no further probes are run and also cancels the captive portal
+    /// task if there were successful probes.  Be sure to only handle this after all the
+    /// required [`ProbeReport`]s have been processed.
+    fn handle_abort_probes(&mut self) {
+        self.outstanding_tasks.probes = false;
+        if self.report.udp {
+            self.outstanding_tasks.captive_task = false;
+        }
+    }
+
+    /// Creates the future which will perform the portmapper task.
+    ///
+    /// The returned future will run the portmapper, if enabled, resolving to it's result.
+    fn prepare_portmapper_task(
+        &mut self,
+    ) -> MaybeFuture<Pin<Box<impl Future<Output = Option<portmapper::ProbeOutput>>>>> {
+        let mut port_mapping = MaybeFuture::default();
+        if !self.skip_external_network {
+            if let Some(port_mapper) = self.port_mapper.clone() {
```

Comment:
I once more find myself rather conflicted.  Part of me agrees that it would be nice to avoid the clone.  Part of me also worries this makes the code harder to understand and easier to get wrong with later refactors:

The `Option` part of `Option<PortMapper` right now only says "you have a portmapper".  Taking it out changes this to "this bit is mutable state that changes depending on where you are in the code.  I know, the methods already has `&mut self` so what am I complaining about.  Still it makes it easier to get right.

If anything I'd rather remove the `Option` altogether: We have no code path which does not pass in the PortMapper.  So why make this optional at all?  Cloning the thing is pretty cheap I think anyway.

Reply:
So I had a go at removing the option, but I don't want to do this for now.  It's a lot of churn, better to be done another time.

If you really feel strongly I can change it into the take, but I'd rather leave it I think until it gets cleaned up a bit better.